### PR TITLE
feat: support aten.as_strided converter

### PR DESCRIPTION
**Pull request by chohk88**
# Description

New feature to support aten.as_strided converter converter. Our implementation focuses on accurately calculating the indices required 

Fixes # ([issue](https://github.com/pytorch/TensorRT/issues/2734))

## Example Usage

Given a tensor `x`:

```python
x = torch.tensor([[1, 2, 3],
                  [4, 5, 6],
                  [7, 8, 9]])
```

#### Example 1: Stride=[1, 2]
- **Operation**: `torch.as_strided(x, (2, 2), (1, 2), 0)`
- **`indices_tensor`**: `[0, 2, 1, 3]` from stride `[1, 2]`.
- **Output**: `[[1, 3], [2, 4]]`

#### Example 2: Stride=[2, 1]
- **Operation**: `torch.as_strided(x, (2, 2), (2, 1), 0)`
- **`indices_tensor`**: `[0, 1, 2, 3]` from stride `[2, 1]`.
- **Output**: `[[1, 2], [3, 4]]`

`as_strided converter` calculates `indices_tensor` using `size` and `stride`, then reshapes the tensor accordingly

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py [hunk 1]
```diff
@@ -259,3 +261,59 @@ def flip(
...
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    input: TRTTensor,
+    size: Sequence[int],
+    stride: Sequence[int],
+    storage_offset: int,
+) -> TRTTensor:
+    assert len(size) == len(stride), "size and stride shapes must be the same"
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py:276]: This can be removed if it is required for Torch to pass. If this condition causes a failure in Torch, it is not required here, since Torch will catch it in the original model when attempting inference.

File: tests/py/dynamo/conversion/test_as_strided_aten.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                (2, 2, 3),
+                (1, 2, 2),
+                1,
+            ),
+            (
+                (200, 200, 200),
+                (9, 9, 3, 2),
+                (2, 2, 2, 3),
+                1,
+            ),
```
Comment (zewenli98) [inline tests/py/dynamo/conversion/test_as_strided_aten.py:41]: Can you add more tests that `size` or `stride` contains 0, and/or the `input` is not a square/cube?

**Review by zewenli98 — CHANGES_REQUESTED** [review]

File: py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py [hunk 3]
```diff
@@ -259,3 +261,59 @@ def flip(
...
+            indices.append(current)
+            return
+        for i in range(size[dim]):  # Recursively compute indices across dimensions
+            nested(
+                rank, size, stride, current + stride[dim] * i, dim + 1
+            )  # Calculate the index for the current dimension and recursively explore further dimensions
+
+    nested(len(size), size, stride, storage_offset, 0)
+
+    indices = torch.tensor(indices, dtype=torch.int)
```
[the remark this answers is not in the document]
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py]: Thank you for your review! I've updated the code based on your suggestions.

File: py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py [hunk 4]
```diff
@@ -259,3 +261,59 @@ def flip(
...
+    # Use gather to reorder elements based on computed indices
+    gather_layer = ctx.net.add_gather(flatten_output, indices_tensor, axis=0)
+    gather_output = gather_layer.get_output(0)
+
+    # Reshape the gathered tensor to the desired size
+    reshape_output = impl.shuffle.reshape(
+        ctx,
+        target,
+        source_ir,
+        f"{name}_reshape",
```
[the remark this answers is not in the document]
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py]: Thank you for your review! I've updated the code based on your suggestions. And also I have changed the name of flatten_output layer.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 5]
```diff
@@ -2133,6 +2133,26 @@ def aten_ops_linear(
...
+) -> Union[TRTTensor, Sequence[TRTTensor]]:
+    return impl.slice.as_strided(
+        ctx,
+        target,
+        source_ir=SourceIR.ATEN,
+        name=name,
+        input=args[0],
+        size=args[1],
+        stride=args[2],
+        storage_offset=args_bounds_check(args, 3, 0),
```
[the remark this answers is not in the document]
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: Thanks for your suggestions! I've made the changes to the code. I also made sure that `storage_offset` is an integer before using it in the `nested` function.

File: py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py [hunk 6]
```diff
@@ -259,3 +261,59 @@ def flip(
...
+        for i in range(size[dim]):  # Recursively compute indices across dimensions
+            nested(
+                rank, size, stride, current + stride[dim] * i, dim + 1
+            )  # Calculate the index for the current dimension and recursively explore further dimensions
+
+    nested(len(size), size, stride, storage_offset, 0)
+
+    indices = torch.tensor(indices, dtype=torch.int)
+
+    indices_tensor = get_trt_tensor(ctx, (indices), f"{name}_indices")
```
[the remark this answers is not in the document]
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py]: Thanks! I have removed unnecessary bracket.

File: tests/py/dynamo/conversion/test_as_strided_aten.py (hunk 2 above)
Comment (zewenli98) [reply tests/py/dynamo/conversion/test_as_strided_aten.py:41]: Alright, it looks like the output doesn't support empty in torch-trt. I think it's ok to add a validator like you have done, but need to remove the duplicated decorator: https://github.com/pytorch/TensorRT/pull/2735/files#diff-9cbf43cfb62cdf682eef77f6fd5cabc488bb5a91ff09bfee5d593827542b2476R2203
